traffic_control/clients/python/trafficops/trafficops/tosession.py

```diff
@@ -45,927 +46,967 @@
 
 # TOSession Class
 class TOSession(restapi.RestApiSession):
-    """
```

Comment:
I understand where you're coming from, but Tabs > spaces. Tabs are for indentation, spaces are for alignment. Also, consider from the first paragraph of PEP8:

> This document gives coding conventions for the Python code comprising the standard library in the main Python distribution.

So it's not only a suggestion, but a suggestion for Python code intended for the standard library.

Reply:
Regardless of whether or not tabs are better than spaces, following a set style guideline makes code easier to read when moving from project to project. In this case, spaces are preferred because they're consistent; a tab can be rendered differently in different views, whereas 4 spaces will always be rendered as 4 spaces. There are other PEP8 style guidelines that make a Python project read like Python, such as `snake_case_function_names` and `CapWordsClassNames`. It's like reading `class my_java_class {...}` in Java; it just doesn't read like Java.

Reply:
+1 on following PEP guidelines. I personally prefer tabs as well, but following the language idioms comes first.

In fact, I'd strongly support adopting a formatter (such as https://github.com/google/yapf) as a standard/requirement for the project, a la `gofmt`.

Reply:
I'd love to see us using something like that as well, especially since we've been adding a lot of Python into the codebase lately. IMO gofmt was a huge win for Go.

Reply:
I have a Pylint configuration file I'll include next time I take a pass at the TO client (it needs to be a real package and not that weird symlink thing).
I disagree that spaces are consistent, tabs are consistent. A tab is one character, always, and always represents an indentation level. Your editor can display them at any desired width per your preferences, without sacrificing the actual contents of the files. That PEP is not meant to force a style guide on anyone using Python, it's specifically for contributors to the reference implementation's standard library.

Reply:
You have to change your editor to choose a particular representation of the tab character, rather than it just being a consistent 4 spaces in any editor you use. For example, when writing the code you might have your editor set to 4 spaces for a tab. When viewing it in `git diff`, it might display as 8 spaces. In a GitHub PR it might be 6 spaces. It's just the style that's been chosen for Python for consistency and readability's sake.

Even though PEP8 mentions it is the style convention to be followed for the standard library, it's generally adopted by big open source Python projects (e.g. Django, OpenStack, Ansible). A major benefit to having a standard style of formatting is that you don't have to argue about the different ways things can be formatted; you just `gofmt` (or equivalent) the code and leave it at that. Generally for Python projects that is PEP8.